gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ui/ClusterStatistics.java

```diff
@@ -0,0 +1,239 @@
+
+      sumHeapStoragePoolUsed += ms.getHeapStoragePoolUsed();
+      sumHeapStoragePoolSize += ms.getHeapStoragePoolSize();
+      sumHeapExecutionPoolUsed += ms.getHeapExecutionPoolUsed();
+      sumHeapExecutionPoolSize += ms.getHeapExecutionPoolSize();
+      sumHeapMemoryUsed += ms.getHeapMemoryUsed();
+      sumHeapMemorySize += ms.getHeapMemorySize();
+
+      sumOffHeapStoragePoolUsed += ms.getOffHeapStoragePoolUsed();
+      sumOffHeapStoragePoolSize += ms.getOffHeapStoragePoolSize();
+      sumOffHeapExecutionPoolUsed += ms.getOffHeapExecutionPoolUsed();
+      sumOffHeapExecutionPoolSize += ms.getOffHeapExecutionPoolSize();
+      sumOffHeapMemoryUsed += ms.getOffHeapMemoryUsed();
+      sumOffHeapMemorySize += ms.getOffHeapMemorySize();
+
+      sumDiskStoreDiskSpace += ms.getDiskStoreDiskSpace();
+
+    }
+
+    sumAggrMemoryUsed = sumHeapMemoryUsed + sumOffHeapMemoryUsed;
+    sumAggrMemorySize = sumHeapMemorySize + sumOffHeapMemorySize;
+
+    this.timeLine.add(lastMemberUpdatedTime);
+
+    this.cpuUsageTrend.add(totalCpuActive / cpuCount);
+
+    this.jvmUsageTrend.add(
+        SnappyUtils.bytesToGivenUnits(sumJvmUsedMemory, SnappyUtils.StorageSizeUnits.GB));
+
+    this.heapStoragePoolUsageTrend.add(
+        SnappyUtils.bytesToGivenUnits(sumHeapStoragePoolUsed, SnappyUtils.StorageSizeUnits.GB));
+
+    this.heapExecutionPoolUsageTrend.add(
+        SnappyUtils.bytesToGivenUnits(sumHeapExecutionPoolUsed, SnappyUtils.StorageSizeUnits.GB));
+
+    this.heapUsageTrend.add(
+        SnappyUtils.bytesToGivenUnits(sumHeapMemoryUsed, SnappyUtils.StorageSizeUnits.GB));
+
+    this.offHeapStoragePoolUsageTrend.add(
+        SnappyUtils.bytesToGivenUnits(sumOffHeapStoragePoolUsed, SnappyUtils.StorageSizeUnits.GB));
+
+    this.offHeapExecutionPoolUsageTrend.add(
+        SnappyUtils.bytesToGivenUnits(sumOffHeapExecutionPoolUsed, SnappyUtils.StorageSizeUnits.GB));
+
+    this.offHeapUsageTrend.add(
+        SnappyUtils.bytesToGivenUnits(sumOffHeapMemoryUsed, SnappyUtils.StorageSizeUnits.GB));
+
+    this.aggrMemoryUsageTrend.add(
+        SnappyUtils.bytesToGivenUnits(sumAggrMemoryUsed, SnappyUtils.StorageSizeUnits.GB));
+
+    this.diskStoreDiskSpaceTrend.add(
+        SnappyUtils.bytesToGivenUnits(sumDiskStoreDiskSpace, SnappyUtils.StorageSizeUnits.GB));
+
+  }
+
+  public Object[] getUsageTrends(int trendType) {
+    Object[] returnArray = null;
+    switch (trendType) {
+      case TREND_TIMELINE:
+        synchronized (this.timeLine) {
+          returnArray = this.timeLine.toArray();
+        }
+        break;
+      case TREND_CPU_USAGE:
+        synchronized (this.cpuUsageTrend) {
+          returnArray = this.cpuUsageTrend.toArray();
+        }
+        break;
+      case TREND_JVM_HEAP_USAGE:
+        synchronized (this.jvmUsageTrend) {
+          returnArray = this.jvmUsageTrend.toArray();
+        }
+        break;
+      case TREND_HEAP_USAGE:
+        synchronized (this.heapUsageTrend) {
```

Comment:
Why synchronize if we are not synchronizing while adding data to queue? We should think about a concurrent structure which is bounded &  non-blocking ( If this method will be executed every time for each browser request. Lets say if multiple users are accessing the dashboard how many times this method gets called ? )
If not simply adding a synchronized keyword while adding to the queue will suffice.

Reply:
As suggested, removed synchronized keyword where present while reading values from queue. Now Synchronizing only when adding values to queue.